cni/pkg/ebpf/ambient_redirect_kern.c

```diff
@@ -0,0 +1,273 @@
+    struct ethhdr  *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+
+    if (data + sizeof(*eth) > data_end) {
+        return TC_ACT_OK;
+    }
+
+    // TODO: support ipv6
+    if (eth->h_proto != htons(ETH_P_IP)) {
+        return TC_ACT_OK;
+    }
+
+    struct iphdr *iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end) {
+        return TC_ACT_OK;
+    }
+    if (iph->protocol != IPPROTO_TCP) {
+        return TC_ACT_OK;
+    }
+    memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+    skb->cb[4] = outbound_cb;
+    dbg("outbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+__section("inbound")
+int inbound_to_ztunnel(struct __sk_buff *skb)
+{
+    if (skb->mark == ZTUNNEL_BYPASS_MARK) {
+        return TC_ACT_OK;
+    }
+
+    void *data = (void *)(long)skb->data;
+    struct ethhdr  *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+    __u32 inbound_cb = INBOUND_CB;
+
+    if (data + sizeof(*eth) > data_end) {
+        return TC_ACT_OK;
+    }
+
+    // TODO: support ipv6
+    if (eth->h_proto != htons(ETH_P_IP)) {
+        return TC_ACT_OK;
+    }
+
+    struct iphdr *iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end) {
+        return TC_ACT_OK;
+    }
+    if (iph->protocol != IPPROTO_TCP) {
+        return TC_ACT_OK;
+    }
+
+    struct ztunnel_info *zi = get_ztunnel_info();
+    if (!zi) {
+        dbg("inbound_to_ztunnel unable to retrieve ztunnel_ifindex\n");
+        return TC_ACT_OK;
+    }
+
+    // overwrite dst_mac
+    memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+    skb->cb[4] = inbound_cb;
+
+    dbg("inbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+__section("ztunnelhostingress")
+int ztunnel_to_app(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    struct ethhdr  *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+
+    if (data + sizeof(*eth) > data_end) {
+        return TC_ACT_OK;
+    }
+
+    // TODO: support ipv6
+    if (eth->h_proto != htons(ETH_P_IP)) {
+        return TC_ACT_OK;
+    }
+    struct iphdr *iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end) {
+        return TC_ACT_OK;
+    }
+    if (iph->protocol != IPPROTO_TCP) {
+        return TC_ACT_OK;
+    }
+
+    skb->mark = ZTUNNEL_BYPASS_MARK;
+    struct app_info *pi = get_app_info_from_ipv4(iph->daddr);
+    if (!pi) {
+        // ip is not in ambient managed mesh
+        // dbg("ztunnelhostingress unable to retrieve pod info: 0x%08x\n", ntohl(iph->daddr));
+        return TC_ACT_OK;
+    }
+
+    // overwrite dst_mac
+    memcpy(eth->h_dest, pi->mac_addr, ETH_ALEN);
+    dbg("ztunnel redirect to app(0x%08x) ifidx: %u\n",
+        ntohl(iph->daddr), pi->ifindex);
+    return bpf_redirect(pi->ifindex, 0);
+}
+
+char __license[] __section("license") = "GPL";
```

Comment:
How feasible is it to make this not depend on any GPL code? Otherwise we are going to need to dimskme investigation on the implications here given GPL is current banned in Istio

Reply:
changed to dual license now